node/heartbeat/monitorEdgeCases_test.go

```diff
@@ -0,0 +1,483 @@
+	heartBeats = mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 10, 0)
+
+	timer.SetSeconds(twentySeconds)
+	//mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 20, 0)
+
+	timer.SetSeconds(hundredSeconds)
+	mon2 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+	//mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	time.Sleep(20 * time.Millisecond)
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 100, 0)
+
+	timer.SetSeconds(hundredFiftySeconds)
+	//mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 150, 0)
+
+	timer.SetSeconds(151)
+	//mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 150, 1)
+
+	timer.SetSeconds(twoHundredSeconds)
+	//mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 150, 50)
+}
+
+// v: |____________________________......
+// o: |___________|.........|___________
+//    0    10    20        100   120   200
+func TestMonitor_ObserverGapValidatorPartlyOnline2(t *testing.T) {
+	t.Parallel()
+
+	storer, _ := storage.NewHeartbeatDbStorer(mock.NewStorerMock(), &mock.MarshalizerFake{})
+
+	timer := &mock.MockTimer{}
+	genesisTime := timer.Now()
+
+	mon1 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+	mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+
+	heartBeats := mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 0, 0)
+
+	timer.SetSeconds(tenSeconds)
+	mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 10, 0)
+
+	timer.SetSeconds(twentySeconds)
+	mon1.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon1.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 20, 0)
+
+	timer.SetSeconds(hundredSeconds)
+	mon2 := createMonitor(storer, genesisTime, unresponsiveDuration, timer)
+	mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 100, 0)
+
+	timer.SetSeconds(120)
+	mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], true, 120, 0)
+
+	timer.SetSeconds(twoHundredSeconds)
+	//mon2.AddHeartbeatMessageToMap(&heartbeat.Heartbeat{Pubkey: []byte(pkValidator)})
+	heartBeats = mon2.GetHeartbeats()
+	verifyHeartBeat(t, heartBeats[0], false, 120, 80)
```

Comment:
I've expected the test to be 
```GO
verifyHeartBeat(t, heartBeats[0], false, 170, 30)
```
because the last heartbeat message was done at timestamp 120. 120 + window( = 50) = 170